sdk/servicebus/Azure.Messaging.ServiceBus/src/ServiceBusException.cs

```diff
@@ -0,0 +1,214 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Globalization;
+using Azure.Messaging.ServiceBus.Receiver;
+
+namespace Azure.Messaging.ServiceBus
+{
+    /// <summary>
+    ///   Serves as a basis for exceptions produced within the Service Bus
+    ///   context.
+    /// </summary>
+    ///
+    /// <seealso cref="System.Exception" />
+    ///
+    public class ServiceBusException : Exception
+    {
+        /// <summary>
+        ///   Indicates whether an exception should be considered transient or final.
+        /// </summary>
+        ///
+        /// <value><c>true</c> if the exception is likely transient; otherwise, <c>false</c>.</value>
+        ///
+        public bool IsTransient { get; }
+
+        /// <summary>
+        ///   The reason for the failure of an Service Bus operation that resulted
+        ///   in the exception.
+        /// </summary>
+        ///
+        public FailureReason Reason { get; }
+
+        /// <summary>
+        ///   The name of the Service Bus to which the exception is associated.
+        /// </summary>
+        ///
+        /// <value>The name of the Service Bus entity, if available; otherwise, <c>null</c>.</value>
+        ///
+        public string EntityName { get; }
+
+        /// <summary>
+        ///   Gets a message that describes the current exception.
+        /// </summary>
+        ///
+        public override string Message
+        {
+            get
+            {
+                if (string.IsNullOrEmpty(EntityName))
+                {
+                    return base.Message;
+                }
+
+                return string.Format(CultureInfo.InvariantCulture, "{0} ({1})", base.Message, EntityName);
+            }
+        }
+
+        /// <summary>
+        ///   Initializes a new instance of the <see cref="ServiceBusException"/> class.
+        /// </summary>
+        ///
+        /// <param name="isTransient"><c>true</c> if the exception should be considered transient; otherwise, <c>false</c>.</param>
+        /// <param name="entityName">The name of the Service Bus entity to which the exception is associated.</param>
+        ///
+        public ServiceBusException(bool isTransient,
```

Comment:
I'd recommend giving the constructors a good hard look.    We collapsed the exception hierarchy late and under time pressure and I made some rushed choices here.   You can probably  consolidate quite a bit with by using optional parameters.

Reply:
https://github.com/Azure/azure-sdk-for-net/issues/9879